Metadata API: Document serialization "repro" issue

jku opened the pull request:
It's not obvious to casual reader that reading metadata and then
writing it might not always produce the same file. It's also not
immediately obvious why this matters.

Document both concepts.

Fixes #1392

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>
---


It's a little wordy but I think it may be useful.

Commit:
Metadata API: Document serialization "repro" issue

It's not obvious to casual reader that reading metadata and then
writing it might not always produce the same file. It's also not
immediately obvious why this matters.

Document both concepts.

Fixes #1392

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

MVrachev reviewed (approved):
The comments look good.

I wonder do we want to add a comment somewhere that even though when deserializing and serializing we don't get byte-for-byte equivalence we do get a consistent signature no matter the whitespaces?

jku commented:
I can see that making the distinction could help... I'll try that (even if the paragraph is a bit long already)

jku commented:
force pushed: added the mention that signatures are guaranteed to stay valid in a deserialize-serialize cycle even if the file content itself is not guaranteed to stay the same.

jku commented:
force pushed: 
* rebased to get recent build fix
* added the mention that signatures are guaranteed to stay valid in a deserialize-serialize cycle even if the file content itself is not guaranteed to stay the same.

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -232,6 +232,13 @@ def to_bytes(
 
+        Note that if bytes are first deserialized into ``Metadata`` and then
+        serialized with ``to_bytes()``, the two are not guaranteed to be the
+        same because of whitespace issues, even if the signatures are
+        guaranteed to stay valid. If byte-for-byte equivalence is required
```
A smart way to mention that signatures are guaranteed to stay valid.

MVrachev reviewed (approved):
LGTM!

lukpueh reviewed (commented):
I like the clarification, modulo two nits:

- the claim *"not guaranteed to be the same"*  depends on the serializer, which is configurable for both `to_file` and `to_bytes`. Maybe it's worth mentioning that?
- the reason *"because of whitespace issues"* is also specific to the default json serializer, and also only part of the problem (e.g. non-guaranteed dict key order is another problem). Maybe it's not necessary to mention a particular reason?

jku commented:
Agreed on both comments, thanks, will update.

jku commented:
I solved this by
* not mentioning whitespace as a reason
* just saying that serialization is not required to be byte-for-byte the same (the implication is that different serializers may make different decision, but that's not said)

lukpueh reviewed (approved):
Thanks for the update, @jku!
